## Scenario

You are a junior data analyst working in the marketing analyst team at Cyclistic, a bike-sharing company in Chicago. Moreno (director of marketing) believes the company’s future success depends on maximizing the number of annual memberships.

Our goal is to design marketing strategies aimed at converting casual riders into annual members. In order to do that, we need to understand how casual riders and annual members use Cyclistic bikes differently.

## About the company

Until now, Cyclistic’s marketing strategy relied on building general awareness and appealing to broad consumer segments. One approach that helped make these things possible was the flexibility of its pricing plans: <b>single-ride passes, full-day passes, and annual memberships.</b> Customers who purchase single-ride or full-day passes are referred to as casual riders. Customers who purchase annual memberships are Cyclistic members.

Cyclistic’s finance analysts have concluded that annual members are much more profitable than casual riders. Although the pricing flexibility helps Cyclistic attract more customers, Moreno believes that <b>maximizing the number of annual members will be key to future growth.</b> Rather than creating a marketing campaign that targets all-new customers, Moreno believes there is a very good chance to convert casual riders into members. She notes that casual riders are already aware of the Cyclistic program and have chosen Cyclistic for their mobility needs.

Moreno has set a clear goal: <b>Design marketing strategies aimed at converting casual riders into annual members.</b> In order to do that, however, the marketing analyst team needs to better understand how annual members and casual riders differ, why casual riders would buy a membership, and how digital media could affect their marketing tactics. Moreno and her team are interested in analyzing the Cyclistic historical bike trip data to identify trends.

# Ask

In this phase, I will identify the main problem that facing by Cyclistic Bike company. I will also understand the needs of Stakeholders.

#### Guiding Quistions

##### What is the Problem you are trying to solve?

##### Problem:

Determine a way to build a profile for annual members and the best marketing strategies to turn casual bike riders into annual members.

##### How can your insights drive business decisions?

The insights will help the marketing team to increase annual members.

#### Key tasks:

1: Identify the business task

2: Consider key Stakeholders


#### Business tasks:

#### These are the questions/business task that would guide the future of the marketing program:

1.	To understand how annual members and casual riders use our Cyclistic bikes differently
2.	Why would casual members upgrade to annual memberships
3.	How can Cyclistic use digital media to influence casual riders to become members?


#### Stakeholders:

1: Lily Moreno, the director of marketing and my manager.

2: Cyclistic executive team.

## Prepare

In this phase, I will collect the Bike data from the below link:</br>
We will be using Cyclistic’s historical trip data [here](https://divvy-tripdata.s3.amazonaws.com/index.html) from 2021 January till 2021 December (202101-divvy-tripdata.zip -> 202112-divvy-tripdata.zip). We will be extracting all of our files into a folder “Divvy_Monthlytripdata” to organize and provide context.

We will also be renaming the files to represent the data more clearly, it would also help with readability for other team members. Below is how I would do so:
(202101-divvy-tripdata.csv) -> (2021_01.csv)
(202102-divvy_tripdata.csv) -> (2021_02.csv)
And so on.

#### Code for Renaming all data files into given names:

In [4]:
import os

file_path = r'E:\Google Data Analytics\Cyclistic Bike Project\Divvy_Monthlytripdata'
list_of_files = os.listdir(file_path)

In [5]:
for file in list_of_files:
    
    base_name, extension = os.path.splitext(file)

    new_name = base_name[:4] + '_' + base_name[4:6] + '.csv'
    
    old_file = os.path.join(file_path, file)
    new_file = os.path.join(file_path, new_name)

    os.rename(old_file, new_file)

#### Guiding Questions:

<b>Where is your data located? </b>

The data is located on this website: [Divvy-tripdata](https://divvy-tripdata.s3.amazonaws.com/index.html).

<b>How is the data organized?</b>

The data is separated by month, each on it's own csv.

<b>Are there issues with bias or credibility in this data? Does your data ROCCC?</b>

Bias isn't a problem, the population of the dataset is it's own clients as bike riders. And have full credibility for the same reason. And finally, it's ROCCC because it's reliable, original, comprehensive, current and cited.

<b>How are you addressing licensing, privacy, security, and accessibility?</b>

The company has their own licence over the dataset. Besides that, the dataset doesn't have any personal information about the riders.

<b>How does it help you answer your question?</b>

It may have some key insights about the riders and their riding style

<b>Are there any problems with the data?</b>

It would be good to have some updated information about the bike stations. Also more information about the riders could be useful.

#### Key tasks

1. Download data and store it appropriately.

2. Identify how it’s organized.

3. Sort and filter the data.

4. Determine the credibility of the data.

#### Deliverable

- A description of all data sources used

The main data source is 12 months (from January 2021 to December 2021) of riding data provided by the Cyclistic Company.

## Process

In this phase, I will clean the data by removing duplicates, handling missing values, identify outlier and make the data accurate for analysis.

#### Import libraries

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px
from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot
init_notebook_mode(connected=True)
import glob

#### Load Data

In [7]:
all_csv_files = [i for i in glob.glob('./Divvy_Monthlytripdata/*.{}'.format('csv'))]

In [8]:
all_csv_files

['./Divvy_Monthlytripdata\\2021_02.csv',
 './Divvy_Monthlytripdata\\2021_03.csv',
 './Divvy_Monthlytripdata\\2021_04.csv',
 './Divvy_Monthlytripdata\\2021_05.csv',
 './Divvy_Monthlytripdata\\2021_06.csv',
 './Divvy_Monthlytripdata\\2021_07.csv',
 './Divvy_Monthlytripdata\\2021_08.csv',
 './Divvy_Monthlytripdata\\2021_09.csv',
 './Divvy_Monthlytripdata\\2021_10.csv',
 './Divvy_Monthlytripdata\\2021_11.csv',
 './Divvy_Monthlytripdata\\2021_12.csv',
 './Divvy_Monthlytripdata\\2021__0.csv']

Reading all CSV Files

In [9]:
all_dataframes = []

In [10]:
for file in all_csv_files:
    
    df = pd.read_csv(file)

    all_dataframes.append(df)


In [11]:
all_dataframes[0].head()

ride_id  rideable_type           started_at             ended_at  \
0  89E7AA6C29227EFF   classic_bike  2021-02-12 16:14:56  2021-02-12 16:21:43   
1  0FEFDE2603568365   classic_bike  2021-02-14 17:52:38  2021-02-14 18:12:09   
2  E6159D746B2DBB91  electric_bike  2021-02-09 19:10:18  2021-02-09 19:19:10   
3  B32D3199F1C2E75B   classic_bike  2021-02-02 17:49:41  2021-02-02 17:54:06   
4  83E463F23575F4BF  electric_bike  2021-02-23 15:07:23  2021-02-23 15:22:37   

         start_station_name start_station_id            end_station_name  \
0  Glenwood Ave & Touhy Ave              525  Sheridan Rd & Columbia Ave   
1  Glenwood Ave & Touhy Ave              525    Bosworth Ave & Howard St   
2        Clark St & Lake St     KA1503000012      State St & Randolph St   
3     Wood St & Chicago Ave              637     Honore St & Division St   
4        State St & 33rd St            13216       Emerald Ave & 31st St   

  end_station_id  start_lat  start_lng    end_lat    end_lng member_casual  
0            660  42.012701 -87.666058  42.004583 -87.661406        member  
1          16806  42.012701 -87.666058  42.019537 -87.669563        casual  
2   TA1305000029  41.885795 -87.631101  41.884866 -87.627498        member  
3   TA1305000034  41.895634 -87.672069  41.903119 -87.673935        member  
4   TA1309000055  41.834733 -87.625827  41.838163 -87.645123        member

#### Fixing data types

In [12]:
# Checking data types of all columns using info method
all_dataframes[0].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49622 entries, 0 to 49621
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ride_id             49622 non-null  object 
 1   rideable_type       49622 non-null  object 
 2   started_at          49622 non-null  object 
 3   ended_at            49622 non-null  object 
 4   start_station_name  45576 non-null  object 
 5   start_station_id    45576 non-null  object 
 6   end_station_name    44264 non-null  object 
 7   end_station_id      44264 non-null  object 
 8   start_lat           49622 non-null  float64
 9   start_lng           49622 non-null  float64
 10  end_lat             49408 non-null  float64
 11  end_lng             49408 non-null  float64
 12  member_casual       49622 non-null  object 
dtypes: float64(4), object(9)
memory usage: 4.9+ MB


First part of the data cleaning process is to fix the data types of all the columns in order to make them easier to manipulate and be more manageable. It should be noted that for several columns the data type was changed to strings, when the data types are displayed, they show up as objects as strings are a type of object in pandas.

In [13]:
for df in all_dataframes:
    
    # started_at
    df['started_at'] = pd.to_datetime(df['started_at'], format="%Y-%m-%d %H:%M:%S")

    # ended_at
    df['ended_at'] = pd.to_datetime(df['ended_at'], format="%Y-%m-%d %H:%M:%S")

    


In [14]:
# Now check the data types where we converted the started_at and ended_at columns into datetime type
all_dataframes[0].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49622 entries, 0 to 49621
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   ride_id             49622 non-null  object        
 1   rideable_type       49622 non-null  object        
 2   started_at          49622 non-null  datetime64[ns]
 3   ended_at            49622 non-null  datetime64[ns]
 4   start_station_name  45576 non-null  object        
 5   start_station_id    45576 non-null  object        
 6   end_station_name    44264 non-null  object        
 7   end_station_id      44264 non-null  object        
 8   start_lat           49622 non-null  float64       
 9   start_lng           49622 non-null  float64       
 10  end_lat             49408 non-null  float64       
 11  end_lng             49408 non-null  float64       
 12  member_casual       49622 non-null  object        
dtypes: datetime64[ns](2), float64(4), object(7)
me

#### Combining Every Dataframe Into One Huge Dataframe

We will combine all the data sets into one massive data set, so all necessary operations can be performed on this dataset.

In [15]:
combined_df = pd.concat(all_dataframes)

In [16]:
combined_df.head()

ride_id  rideable_type          started_at            ended_at  \
0  89E7AA6C29227EFF   classic_bike 2021-02-12 16:14:56 2021-02-12 16:21:43   
1  0FEFDE2603568365   classic_bike 2021-02-14 17:52:38 2021-02-14 18:12:09   
2  E6159D746B2DBB91  electric_bike 2021-02-09 19:10:18 2021-02-09 19:19:10   
3  B32D3199F1C2E75B   classic_bike 2021-02-02 17:49:41 2021-02-02 17:54:06   
4  83E463F23575F4BF  electric_bike 2021-02-23 15:07:23 2021-02-23 15:22:37   

         start_station_name start_station_id            end_station_name  \
0  Glenwood Ave & Touhy Ave              525  Sheridan Rd & Columbia Ave   
1  Glenwood Ave & Touhy Ave              525    Bosworth Ave & Howard St   
2        Clark St & Lake St     KA1503000012      State St & Randolph St   
3     Wood St & Chicago Ave              637     Honore St & Division St   
4        State St & 33rd St            13216       Emerald Ave & 31st St   

  end_station_id  start_lat  start_lng    end_lat    end_lng member_casual  
0            660  42.012701 -87.666058  42.004583 -87.661406        member  
1          16806  42.012701 -87.666058  42.019537 -87.669563        casual  
2   TA1305000029  41.885795 -87.631101  41.884866 -87.627498        member  
3   TA1305000034  41.895634 -87.672069  41.903119 -87.673935        member  
4   TA1309000055  41.834733 -87.625827  41.838163 -87.645123        member

In [ ]:
# Now check the shape of the combined dataset

combined_df.shape

(5595063, 13)

#### Feature Creation

After going through our business task, we need to add more columns for detail analysis:

<b>Day of the week</b>

In [17]:
combined_df['day_of_week'] = combined_df['started_at'].dt.strftime('%a')
# Day order
day_order = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]
# Convert 'day_of_week' column to categorical with custom order
cat_dtype = pd.CategoricalDtype(categories=day_order, ordered=True)
combined_df['day_of_week'] = combined_df['day_of_week'].astype(cat_dtype)

<b>Start hour</b>

In [18]:
combined_df['starting_hour'] = combined_df['started_at'].dt.strftime('%H')

<b>Month</b>

In [19]:
combined_df['month'] = combined_df['started_at'].dt.strftime('%m')

<b>Trip Duration</b>

In [20]:
combined_df['trip_duration'] = (combined_df['ended_at'] - combined_df['started_at'])/pd.Timedelta(minutes=1)

#### Remove those records which trip_duration less than or equal to 0

In [21]:
negative_record = combined_df[combined_df['trip_duration']<=1]

In [22]:
negative_record.shape

(86111, 17)

In [23]:
new_df = combined_df[combined_df['trip_duration']>1].reset_index().drop(columns='index')

In [24]:
new_df.shape

(5508952, 17)

#### Identifying missing values

In [25]:
round((new_df.isnull().sum()/len(new_df))*100,2)

ride_id                0.00
rideable_type          0.00
started_at             0.00
ended_at               0.00
start_station_name    12.17
start_station_id      12.17
end_station_name      12.97
end_station_id        12.97
start_lat              0.00
start_lng              0.00
end_lat                0.09
end_lng                0.09
member_casual          0.00
day_of_week            0.00
starting_hour          0.00
month                  0.00
trip_duration          0.00
dtype: float64

In [26]:
df_without_nulls = new_df.dropna()

In [27]:
df_without_nulls.shape

(4528291, 17)

#### Handling duplicates

In [28]:
duplicate_records = df_without_nulls[df_without_nulls.duplicated('ride_id')]

In [29]:
duplicate_records.head()

Empty DataFrame
Columns: [ride_id, rideable_type, started_at, ended_at, start_station_name, start_station_id, end_station_name, end_station_id, start_lat, start_lng, end_lat, end_lng, member_casual, day_of_week, starting_hour, month, trip_duration]
Index: []

There are no duplicate records found in the dataset.

In [31]:
df_without_nulls.shape

(4528291, 17)

#### Handling outliers

In [32]:
pd.set_option('display.float_format', lambda x: '%.2f' % x)

In [33]:
df_without_nulls.describe()

started_at                       ended_at  \
count                        4528291                        4528291   
mean   2021-07-25 05:23:27.805043968  2021-07-25 05:45:33.472131840   
min              2021-01-01 00:02:24            2021-01-01 00:08:39   
25%       2021-06-04 17:14:53.500000     2021-06-04 17:36:03.500000   
50%              2021-07-28 17:30:05            2021-07-28 17:48:00   
75%       2021-09-18 15:13:12.500000            2021-09-18 15:40:43   
max              2021-12-31 23:59:48            2022-01-03 17:32:18   
std                              NaN                            NaN   

       start_lat  start_lng    end_lat    end_lng  trip_duration  
count 4528291.00 4528291.00 4528291.00 4528291.00     4528291.00  
mean       41.90     -87.64      41.90     -87.64          22.09  
min        41.65     -87.83      41.65     -87.83           1.02  
25%        41.88     -87.66      41.88     -87.66           7.13  
50%        41.90     -87.64      41.90     -87.64          12.38  
75%        41.93     -87.63      41.93     -87.63          22.32  
max        42.06     -87.53      42.17     -87.52       55944.15  
std         0.04       0.02       0.04       0.02         185.63

Now you can see the trip_duration column which have maximum 55944 (minutes) value which is extream value and it will effect our analysis.

In [6]:
import numpy as np

trip_duration = df_without_nulls['trip_duration']

ventiles = np.percentile(trip_duration, np.arange(0, 101, 5))
ventiles = np.around(ventiles, decimals=3)

output = ""

for i, ventile in enumerate(ventiles):
    output += f"{i * 5}%: {ventile} "

print(output)

0%: 1.017 5%: 3.3 10%: 4.4 15%: 5.333 20%: 6.233 25%: 7.133 30%: 8.05 35%: 9.017 40%: 10.05 45%: 11.15 50%: 12.383 55%: 13.767 60%: 15.35 65%: 17.233 70%: 19.5 75%: 22.317 80%: 25.933 85%: 30.817 90%: 38.833 95%: 58.317 100%: 55944.15 


In [7]:
58.317 - 1.017

57.3

The difference between 1% and 95% is 57.3 minutes. Because of that, in the analysis of this variable we are going to use a subset of the dataset without outliners. The subset will contain 95% of the dataset.

In [8]:
lower_percentile = np.percentile(df_without_nulls['trip_duration'], 1)
upper_percentile = np.percentile(df_without_nulls['trip_duration'], 95)

df_without_outlier = df_without_nulls[
    (df_without_nulls['trip_duration']>lower_percentile)&
    (df_without_nulls['trip_duration']<upper_percentile)]


In [9]:
pd.set_option('display.float_format', lambda x: '%.2f' % x)

In [10]:
df_without_outlier.describe()

start_lat  start_lng    end_lat    end_lng  starting_hour      month  \
count 4255671.00 4255671.00 4255671.00 4255671.00     4255671.00 4255671.00   
mean       41.90     -87.64      41.90     -87.64          14.25       7.31   
std         0.04       0.02       0.04       0.02           5.02       2.49   
min        41.65     -87.83      41.65     -87.83           0.00       1.00   
25%        41.88     -87.66      41.88     -87.66          11.00       6.00   
50%        41.90     -87.64      41.90     -87.64          15.00       7.00   
75%        41.93     -87.63      41.93     -87.63          18.00       9.00   
max        42.06     -87.53      42.08     -87.53          23.00      12.00   

       trip_duration  
count     4255671.00  
mean           15.30  
std            11.25  
min             1.92  
25%             7.03  
50%            11.87  
75%            20.28  
max            58.30

In [11]:
print("Removed", len(df_without_nulls) - len(df_without_outlier), "rows as outliers")

Removed 272620 rows as outliers


In [12]:
df_without_outlier.shape

(4255671, 17)

#### Save the Clean data file

In [13]:
df_without_outlier.to_csv('clean_data.csv', index=False)

#### Load the Clean data file

In [2]:
new_df = pd.read_csv('clean_data.csv')

In [3]:
new_df.shape

(4255671, 17)

#### Guiding questions

<b>What tools are you choosing and why?</b>

I'm using Python for this project, for two main reasons: Because of the large dataset and to gather experience with the language.

<b>Have you ensured your data’s integrity?</b>

Yes, the data is consistent throughout the columns.

<b>What steps have you taken to ensure that your data is clean?</b>

First the null values where removed, then the check the duplicate values but not found any duplicate values then remove outlier values.

<b>Have you documented your cleaning process so you can review and share those results?</b>

Yes, it's all documented in this Python notebook.

#### Key tasks

1. Check the data for errors.
2. Choose your tools.
3. Transform the data so you can work with it efectively
4. Document the cleaning process

#### Deliverable

 - Documentation of any cleaning or manipulation of data

### Analyze

Now it’s time to analyze the data and look for key information that we can perform analysis on.

As mentioned just a moment ago, it is imperative to always remind yourself of the business task at hand during this stage. In order to answer our first business question, it would be beneficial to plot a few of our observations revolving around:

1. How much of the data is about members and how much is about casuals?
2. How do casual and members use their bikes differently throughout the week
3. Peak hours of bike usage between casual and annual members
4. Bike usage by each month
5. The average trip duration between casual and annual members
6. Rideable Type usage by Member Type

<b>1: How much of the data is about members and how much is about casuals?</b>

In [24]:
pivot_table = pd.pivot_table(new_df,
                             index='member_casual',
                             values='ride_id',
                             aggfunc=[len, lambda x: (len(x) / len(new_df)) * 100],
                             margins=True,
                             margins_name='Total Count')

pivot_table.columns = ['Count', 'Percentage']

print(pivot_table)

                 Count  Percentage
member_casual                     
casual         1806265       42.44
member         2449406       57.56
Total Count    4255671      100.00


In [27]:
casual_members_data = new_df.groupby(['member_casual'])['ride_id'].count().reset_index()

In [28]:
casual_members_data.head()

member_casual  ride_id
0        casual  1806265
1        member  2449406

In [30]:
px.bar(casual_members_data, y='member_casual', x='ride_id',
       title='Number of Rides by Member Type',
        color = 'member_casual', 
        height = 300,
        text = 'ride_id', 
        labels = {'ride_id': 'No. of Rides', 'member_casual': 'Member/Casual'},
        hover_name = 'member_casual', hover_data = {'member_casual': False, 'ride_id': True}, 
        color_discrete_map = {'casual': '#FF934F', 'member': '#058ED9'})


Observation:

As we can see on the member x casual table, members have a bigger proporcion of the dataset, composing ~57%, ~12% bigger than the count of casual riders.

<b>2: How do casual and members use their bikes differently throughout the week</b>

In [4]:
# Day order
day_order = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]
# Convert 'day_of_week' column to categorical with custom order
cat_dtype = pd.CategoricalDtype(categories=day_order, ordered=True)
new_df['day_of_week'] = new_df['day_of_week'].astype(cat_dtype)

In [5]:
pd.pivot_table(new_df, index=['day_of_week', 'member_casual'],
               values='ride_id',
               aggfunc='count',
               margins=True,
               margins_name='Total Count')

ride_id
day_of_week member_casual         
Mon         casual          200128
            member          334202
Tue         casual          192503
            member          375258
Wed         casual          197512
            member          384706
Thu         casual          204033
            member          361121
Fri         casual          260538
            member          353011
Sat         casual          407761
            member          342906
Sun         casual          343790
            member          298202
Total Count                4255671

In [45]:
day_of_week = new_df.groupby(['day_of_week','member_casual'])['ride_id'].count().reset_index()

In [48]:
px.line(day_of_week, x='day_of_week', y='ride_id',
        title='Riders usage by Member type in Whole Week',
        color='member_casual',
        line_shape='spline',
        markers=True,
        labels={'ride_id': 'No. of Rides', 'day_of_week': 'Week Days', 'member_casual': 'Member/Casual'},
        hover_name='member_casual', hover_data={'member_casual': False,'day_of_week':True, 'ride_id': True},
        color_discrete_map={'casual': '#FF934F', 'member': '#058ED9'},
        category_orders={'day_of_week': day_order}
       )

Observation:

Here, we can find that overall ridership for annual members is fairly stable across the week. Such would indicate that there is a possibility that annual members are using the bikes as their main mode of transportation.

On the other hand, ridership for casual members is fairly low on the weekdays but starts to ramp up on Fridays and eventually peaks on Saturdays.

<b>3: Peak hours of bike usage between casual and annual members</b>

We first see the Peak hourse of bike usage between casual and annual members on the weekdays.

In [50]:
weekday_names = ["Mon", "Tue", "Wed", "Thu", "Fri"]  # List of weekday names

# Filter the dataframe to keep only the weekday records
weekday_df = new_df[new_df['day_of_week'].isin(weekday_names)]

In [54]:
peak_hours_weekdays = weekday_df.groupby(['starting_hour', 'member_casual'], as_index=False).count()

In [55]:
peak_hour_fig = px.line(peak_hours_weekdays, x='starting_hour', y='ride_id',
                 title='Hourly usage of bikes on the weekdays',
                 color='member_casual',
                 line_shape='spline',
                 markers=True,
                 labels={'ride_id':'No. of Rides', 'starting_hour':'24 Hours', 'member_casual':'Member/Casual'},
                 hover_name='member_casual', hover_data={'member_casual':False, 'month':False, 'ride_id': True},
                 color_discrete_map={'casual':'#FF934F', 'member':'#058ED9'},
                #  category_orders={'starting_hour': hour_order
                #                   }
                )

peak_hour_fig.update_xaxes(range=[0, 23], dtick=1)
peak_hour_fig.show()

Observation:

Based on the weekday graph, it would further reinforce my previous hypothesis whereby annual members are working adults, as we can see from:

- 7 am-8 am: A rally in usage, which could indicate when they’ve begun commuting to work
- 12 pm: An increase in usage, which would indicate lunch hour
- 5 pm: A peak in usage, which again falls in line with the office off-hours

Now see the Peak hourse of bike usage between casual and annual members on the weekendays.

In [9]:
weekendays = ["Sat", "Sun"]

weekend_df = new_df[new_df['day_of_week'].isin(weekendays)]

In [14]:
group_weekendays = weekend_df.groupby(['starting_hour', 'member_casual'], as_index=False).count()

In [16]:
weenkend_fig = px.line(group_weekendays, x='starting_hour', y='ride_id',
                       color='member_casual',
                       title='Hourly usage of bikes on the weekends',
                       line_shape='spline',
                       markers=True,
                       labels={'ride_id':'No. of Riders', 'member_casual': 'Member/Casual', 'starting_hour':'24 Hours'},
                       hover_name='member_casual', hover_data={'member_casual':False, 'ride_id':True},
                       color_discrete_map={'casual':'#FF934F', 'member':'#058ED9'})

weenkend_fig.update_xaxes(range=(0,23), dtick=1)
weenkend_fig.show()

Observation:

The weekends, on the other hand, see a dramatic increase in ridership for the casual members starting from 11 am and peaks in the afternoon.

<b>4: Bike usage by each month </b>

Steps:

1. Need a line chart that show bike usage by each month
2. on x-axis months will be shown.
3. on y-axis ride_id will be shown.
4. legend will be member_casual

In [18]:
pd.pivot_table(new_df, index=['month','member_casual'],
               values='ride_id',
               aggfunc='count',
               margins=True,
               margins_name='Total Count')

ride_id
month       member_casual         
1           casual           13626
            member           66912
2           casual            7553
            member           33088
3           casual           64730
            member          125891
4           casual          102853
            member          171462
5           casual          183749
            member          225650
6           casual          263459
            member          293846
7           casual          325065
            member          311214
8           casual          303882
            member          321106
9           casual          262528
            member          316887
10          casual          171605
            member          278164
11          casual           65139
            member          178846
12          casual           42076
            member          126340
Total Count                4255671

In [19]:
month_group = new_df.groupby(['month', 'member_casual'], as_index=False).count()

In [20]:
month_fig = px.line(month_group, x='month', y='ride_id',
                    color='member_casual',
                    title='Bike usage by Member Type in each Month',
                    line_shape='spline',
                    markers=True,
                    labels={'ride_id':'No. of Riders', 'member_casual':'Member/Casual', 'month': 'Months (Jan 2021 - Dec 2021)'},
                    hover_name='member_casual', hover_data={'member_casual':False, 'ride_id':True, 'month':True},
                    color_discrete_map={'casual':'#FF934F', 'member':'#058ED9'})

month_fig.update_xaxes(range=[1,12], dtick=1)
month_fig.show()

Observation:

As we can see, ridership starts to freefall during the later months of the year. This is possible due to the change of seasons as usually in the months of October, the temperature starts to drop and the possibility of snow entails after.

Interestingly, casual member ridership peaks in the month of July. Knowing that schools in Chicago end around the middle of July, and resumes at end of August, we could hypothesize that the majority of the said users are not high school students.

<b>5: The average trip duration between casual and annual members</b>

Goal:

- Display horizontal Bar chart which show average trip duration of Member and Casual

Steps:

- On x-axis rider_id will be show
- On y-axis member_casual will be show
- legend will be member_casual

In [25]:
pd.pivot_table(new_df, index=['day_of_week','member_casual'],
               values='trip_duration',
               aggfunc='mean',
               margins=True,
               margins_name='Total Average')

trip_duration
day_of_week   member_casual               
Mon           casual             18.796741
              member             12.414479
Tue           casual             17.373827
              member             12.165212
Wed           casual             17.072825
              member             12.216255
Thu           casual             16.859338
              member             12.114794
Fri           casual             17.970157
              member             12.479820
Sat           casual             20.009151
              member             14.108656
Sun           casual             20.485957
              member             14.371536
Total Average                    15.301848

In [32]:
average_group = new_df.groupby(['day_of_week', 'member_casual'], as_index=False)['trip_duration'].mean()
average_group['trip_duration'] = average_group['trip_duration'].round(2)

In [33]:
average_fig = px.bar(average_group, x='day_of_week', y='trip_duration',
                     color='member_casual',
                     title='Average Trip duration usage by Member Type in whole week',
                     labels={'trip_duration':'Average Trip Duration', 'member_casual':'Member/Casual', 'day_of_week':'Week Days'},
                     hover_name='member_casual', hover_data={'member_casual':False, 'day_of_week':True, 'trip_duration':True},
                     color_discrete_map={'casual':'#FF934F', 'member':'#058ED9'}
                     )

average_fig.show()

Observation:

Here, we see that casual members have significantly longer average trip durations than annual members, nearly double in fact. This could signify that casual members are mainly using the bikes for leisure and or possibly sports activities.

Whereas annual members would very likely use the bikes to commute from their living quarters to their offices and vice versa.

<b>6: Rideable Type usage by Member Type</b>

In [36]:
pd.pivot_table(new_df, index=['rideable_type', 'member_casual'], values='ride_id',
               aggfunc=['count'],
               margins=True,
               margins_name='Total Count')

count
                             ride_id
rideable_type member_casual         
classic_bike  casual         1132680
              member         1913084
docked_bike   casual          234704
              member               1
electric_bike casual          438881
              member          536321
Total Count                  4255671

In [5]:
rideable_type_group = new_df.groupby(['rideable_type', 'member_casual']).agg(
    count=('ride_id', 'size'),
    percent = ('ride_id',lambda x: (len(x)/len(new_df))*100),
    ).reset_index()

rideable_type_group['percent'] = rideable_type_group['percent'].round(2).astype(str) + '%'

In [6]:
rideable_type_group.head()

rideable_type member_casual    count percent
0   classic_bike        casual  1132680  26.62%
1   classic_bike        member  1913084  44.95%
2    docked_bike        casual   234704   5.52%
3    docked_bike        member        1    0.0%
4  electric_bike        casual   438881  10.31%

In [10]:
rideable_type_fig = px.bar(rideable_type_group, y='rideable_type', x='count',
                           color='member_casual',
                           title='Rideable Type usage by Member Type',
                           labels={'count': 'Count', 'rideable_type': 'Rideable Type', 'member_casual':'Member/Casual'},
                           hover_name='member_casual', hover_data={'member_casual':False, 'percent':True, 'count':True},
                           color_discrete_map={'casual':'#FF934F', 'member':'#058ED9'}
                           )

rideable_type_fig.show()

Observations:

- Members have a bigger preference for classic bikes and electric bikes which show ~44% for classic_bike and ~12% for electric_bike.
- Casual have ~26% preference for classic bikes and 16% for electric_bike.
- Casual also use the docked_bike by ~5% but Member dont use the docked_bike.

#### Guiding questions

- <b>How should you organize your data to perform analysis on it?</b>

The data has been organized into a single CSV concatenating all the files from the dataset.

- <b>Has your data been properly formatted?</b>

Yes, all the columns have their correct data type.

- <b>What surprises did you discover in the data?</b>

One of the main surprises is how members differ from casuals when analysed from weekdays. Also that members have less riding time than casual.

- <b>What trends or relationships did you find in the data?</b>

- There are more members than casuals in the dataset.
- There are more data points in the July 2021.
- There are more of a difference between the flow of members/casual from midweek to weekends.
- Members use bikes on schedules that differs from casual.
- Members have less riding time.
- Members tend to prefer classic bikes.


- <b>How will these insights help answer your business questions?</b>

This insights helps to build a profile for members.

#### Key tasks

1. Aggregate your data so it’s useful and accessible.
2. Organize and format your data.
3. Perform calculations.
4. Identify trends and relationships.

#### Deliverable

A summary of your analysis

### Share

Let's go through the main finds and try to arrive at a conclusion.

What we know about the dataset:

- Members have the biggest proportion of the dataset, ~12% bigger thand casuals.
- The month with the biggest count of data points was July.
- In all months we have more members' rides than casual rides.
- The difference of proporcion of member x casual is smaller in the last semester of 2021.
- Temperature heavily influences the volume of rides in the month.
- The biggest volume of data is on the the weekend.
- There's a bigger volume of bikers in the afternoon.

It's possible to notice that the distribution of rides by month is cyclical through years, it's influenced by the temperature. The remaining question is: Why are there more members than casual? One plausible answer is that members have a bigger need for the bikes than casuals, as can be seen on how there are more members than casuals on cold months.

Besides that, we have more bike rides on the weekends. Maybe because on those days the bikes were utilized for more recreational ways. This even more plausible when knowing that There's a bigger volume of bikers in the afternoon.

Now for how members differs from casuals:

- Members may have the biggest volume of data, besides on saturday. On this weekday, casuals take place as having the most data points.
- Weekends have the biggest volume of casuals, starting on friday, a ~20% increase.
- We have more members during the morning, mainly between 5am and 11am.
- There's a big increase of data points in the midweek between 6am to 8am for members. Then it fell a bit. Another big increase is from 4pm to 5pm.
- During the weekend we have a bigger flow of casuals between 11am to 4pm.
- Members have a bigger preference for classic bikes, 44% more.
- Casuals have more riding time than members.
- Riding time for members keeps unchanged during the midweek, increasing during weekends.
- Casuals follow a more curve distribution, peaking on sundays and valleying on wednesday/thursday.

What we can take from this information is that members have a more fixed use for bikes besides casuals. Their uses is for more routine activities, like:

- Go to work.
- Use it as an exercise

This can be proven we state that we have more members in between 6am to 8am and at 5pm to 6pm. Also, members may have set routes when using the bikes, as proven by riding time for members keeps unchanged during the midweek, increasing during weekends. The bikes is also heavily used for recreation on the weekends, when riding time increases and casuals take place.

Members also have a bigger preference for classic bikes, so they can exercise when going to work.

Concluding:

- Members use the bikes for fixed activities, one of those is going to work.
- Bikes are used for recreation on the weekends.
- Rides are influenced by temperature.

#### Guiding questions

- <b>Were you able to answer the question of how annual members and casual riders use Cyclistic bikes differently?</b>

Yes. The data points to several differences between casuals and members.

- <b>What story does your data tell?</b>

The main story the data tells is that members have set schedules, as seen timestamps on weekdays chart. Those timestamps point out that members use the bikes for routine activities, like going to work. Average time Chart also point out that they have less riding time, because they have a set route to take.

- <b>How do your findings relate to your original question?</b>

The findings build a profile for members, relating to "Find the keys differences between casuals and annual riders", also knowing why they use the bikes helps to find "How digital media could influence them".

- <b>Who is your stakeholders? What is the best way to communicate with them?</b>

The main target stakeholders are my cyclistic marketing analytics team and Lily Moreno. The best way to communicate is through a slide presentation of the findings.

- <b>Can data visualization help you share your findings?</b>

Yes, the main core of the finds is through data visualization.

- <b>Is your presentation accessible to your stakeholders?</b>

Yes, the plots were made using vibrant colors, and corresponding labels.

#### Key tasks

1. Determine the best way to share your findings.
2. Create effective data visualizations.
3. Present your findings.
4. Ensure your work is accessible.

#### Deliverable

- Supporting visualizations and key findings

### Act

The act phase would be done by the marketing team of the company. The main takeaway will be the top three recommendations for the marketing.

#### Guiding questions

- <b>What is your final conclusion based on your analysis?</b>

Members and casual have different habits when using the bikes. The conclusion is further stated on the share phase.

- <b>How could your team and business apply your insights?</b>

The insights could be implemented when preparing a marketing campaign for turning casual into members. The marketing can have a focus on workers as a green way to get to work.

- <b>What next steps would you or your stakeholders take based on your findings?</b>

Further analysis could be done to improve the findings, besides that, the marketing team can take the main information to build a marketing campaign.

- <b>Is there additional data you could use to expand on your findings?</b>

1. Mobility data.
2. Improved climate data.
3. More information members.

#### Deliverable

- Your top three recommendations based on your analysis

1. Build a marketing campaign focusing on show how bikes help people to get to work, while maintaining the planet green and avoid traffic. The ads could be show on professional social networks.
2. Increase benefits for riding during cold months. Coupons and discounts could be handed out.
3. As the bikes are also used for recreations on the weekends, ads campaigns could also be made showing people using the bikes for exercise during the weeks. The ads could focus on how practical and consistent the bikes can be.